skimage/metrics/overlap.py

```diff
@@ -0,0 +1,198 @@
+
+    @property
+    def ndim(self):
+        return len(self.top_left)
+
+    def __bool__(self):
+        return bool(self.area > 0)  # cast needed to avoid np.bool_
+
+    def __eq__(self, other: Rectangle):
+        """Return true if 2 rectangles have the same position and dimension."""
+        if not isinstance(other, Rectangle):
+            raise TypeError(
+                    'Equality can only be checked with another Rectangle'
+                    )
+
+        return (np.all(self.top_left == other.top_left)
+                and np.all(self.bottom_right == other.bottom_right))
+
+    def __repr__(self):
+        return (f'Rectangle({tuple(self.top_left)}, '
+                f'bottom_right={tuple(self.bottom_right)})')
+
+    def __str__(self):
+        return self.__repr__()
+
+    @property
+    def area(self):
+        """Return the rectangle area in pixels."""
+        return np.prod(self.dimensions)
+
+    @property
+    def dimensions(self):
+        """Return the (height, width) dimensions in pixels."""
+        return self.bottom_right - self.top_left
+
+
+def _disjoint(rectangle1, rectangle2):
+    """Check whether two rectangles are disjoint
+
+    Adapted from post from Aman Gupta [1]_.
+
+    Parameters
+    ----------
+    rectangle1, rectangle2 : Rectangle
+        Input rectangles.
+
+    Returns
+    -------
+    disjoint : bool
+        True if the rectangles don't share any points.
+
+    References
+    ----------
+    .. [1] https://www.geeksforgeeks.org/find-two-rectangles-overlap/
+    """
+    disjoint = (
+            np.any(rectangle1.bottom_right < rectangle2.top_left)
+            or np.any(rectangle2.bottom_right < rectangle1.top_left)
+            )
+    return disjoint
+
+
+def intersect(rectangle1, rectangle2):
+    """Return the Rectangle corresponding to the intersection of the inputs.
+
+    If the two inputs are disjoint, the returned rectangle is the "null"
+    rectangle, with top left corner at (0, 0) and dimensions (0, 0).
+
+    Parameters
+    ----------
+    rectangle1, rectangle2 : Rectangle
+        Input rectangles.
+
+    Returns
+    -------
+    intersected : Rectangle
+        The rectangle produced by intersecting ``rectangle1`` and
+        ``rectangle2``.
+
+    Examples
+    --------
+    >>> r0 = Rectangle((0, 0), bottom_right=(2, 3))
+    >>> r1 = Rectangle((1, 2), bottom_right=(4, 4))
+    >>> intersect(r0, r1)
+    Rectangle((1, 2), bottom_right=(2, 3))
+
+    Since 0-area rectangles evaluate to False, this function can be used to
+    test for disjoint rectangles.
+
+    >>> r2 = Rectangle((10, 10), dimensions=(3, 3))
+    >>> if not intersect(r1, r2):
+    ...     print('r1 and r2 are disjoint')
+    r1 and r2 are disjoint
+    """
+    if _disjoint(rectangle1, rectangle2):
+        # return the "null rectangle" if they are disjoint
+        ndim = rectangle1.ndim
+        return Rectangle((0,) * ndim, bottom_right=(0,) * ndim)
```

Comment:
This convention may give some counter intuitive logic!
```python
r1 = Rectangle((0, 0), bottom_right=(1, 1))
r2 = Rectangle((2, 2), bottom_right=(3, 3))

r1_r2_intersection = intersect(r1, r2)  # Rectangle((0, 0), (0, 0))

intersect(r1, r2) is False
intersect(r2, r1_r2_intersection) is False
intersect(r1, r1_r2_intersection) is True #!!!!
```

Reply:
I don't get that result @rfezzani?

```python
In [1]: from skimage.metrics import Rectangle, intersect

In [2]: r1 = Rectangle((0, 0), bottom_right=(1, 1))

In [3]: r2 = Rectangle((2, 2), bottom_right=(3, 3))

In [4]: r12 = intersect(r1, r2)

In [5]: r12
Out[5]: Rectangle((0, 0), bottom_right=(0, 0))

In [6]: bool(r12)
Out[6]: False

In [7]: bool(intersect(r1, r2))
Out[7]: False

In [8]: bool(intersect(r2, r12))
Out[8]: False

In [9]: bool(intersect(r1, r12))
Out[9]: False
```

intersections are only true if they have more than 0 area, which R((0, 0), (0, 0)) does not.

Reply:
Huh... Honestly, I did not ran the code, but I expected that the corner of a rectangle is included in it... Isn't it what we want?

Reply:
I think @rfezzani is having a good point. 
To me it also feels that eventhough `Rectangle((0, 0), bottom_right=(0, 0))` has a 0 area it should __intersect__ with any rectangle starting at top_left (0,0).

However the overlap would indeed  be 0.

Shapely actually distinguish intersection (any point in common) and [overlap ](https://shapely.readthedocs.io/en/latest/manual.html#object.overlaps) (more than one point).
I think it would be good to have a similar approach for consistency of the python language.
This [reference](https://giswiki.hsr.ch/images/3/3d/9dem_springer.pdf) (page 3) of shapely doc seems to provide rock-solid definitions for these relations.

So should we rename the current `intersect` to `overlap` ?
Do we need a shapely's like `intersect` ? as far as I am concerned I would not need it.

Reply:
Thank you @LauLauThom for these inverstigations. I like the idea of using `overlap` instead of `intersect`.
What if the `overlap` function returns the overlap region area instead of a `Rectangle` object as `intersect` does?
This way, there is no need for defining `Rectangle.__bool__` and the syntax
```python
if overlap(r1, r2):
    ...
```
Is still valid.

Reply:
I think by returning a Rectangle we thought of being more generic, like for debugging you could actually see the intersection rectangle.
But since it's hard to find an intuitive and consistent solution for non-overlapping pair of rectangles (I though of returning None instead of a null rectangle otherwise), I would be totally fine with this simpler solution of directly returning the area.
What do you think @jni ?